src/core/media/mpris/media_player2.cpp

```diff
@@ -0,0 +1,367 @@
+    qint64 duration() const { return m_player ? m_player->duration() : 0; }
+    qint16 backend() const { return m_player ? m_player->backend() : 0; }
+    qint16 orientation() const { return m_player ? m_player->orientation() : 0; }
+    qint16 lifetime() const { return m_player ? m_player->lifetime() : 0; }
+    qint16 audioStreamRole() const { return m_player ? m_player->audioStreamRole() : 0; }
+
+public Q_SLOTS:
+    void Next() { if (m_player) m_player->next(); }
+    void Previous() { if (m_player) m_player->previous(); }
+    void Pause() { if (m_player) m_player->pause(); }
+    void PlayPause();
+    void Stop() { if (m_player) m_player->stop(); }
+    void Play() { if (m_player) m_player->play(); }
+    void Seek(quint64 microSeconds);
+    void SetPosition(const QDBusObjectPath &, quint64) {} // Never implemented
+    /* The OpenUri should not return anything, but since the previous
+     * implementation was returning a boolean, let's keep doing that. */
+    void OpenUri(const QDBusMessage &);
+
+Q_SIGNALS:
+    void Seeked(quint64 microSeconds);
+    void AboutToFinish();
+    void EndOfStream();
+    void PlaybackStatusChanged(quint16 status); // TODO: remove, we have PropertiesChanged already
+    void VideoDimensionChanged(quint32 height, quint32 width);
+    void Error(qint16 code);
+    void Buffering(int percent); // TODO: set a fixed type
+
+private:
+    media::PlayerImplementation *m_player;
+    DBusPropertyNotifier m_propertyNotifier;
+};
+
+class MediaPlayer2Private {
+public:
+    MediaPlayer2Private(MediaPlayer2 *q);
+
+private:
+    friend class MediaPlayer2;
+    QDBusConnection m_connection;
+    PlayerAdaptor *m_playerAdaptor;
+};
+
+} // namespace mpris
+
+PlayerAdaptor::PlayerAdaptor(const QDBusConnection &connection,
+                             QObject *parent):
+    QDBusAbstractAdaptor(parent),
+    m_player(nullptr),
+    m_propertyNotifier(connection, objectPath, this)
+{
+}
+
+void PlayerAdaptor::setPlayer(media::PlayerImplementation *impl)
+{
+    using PlayerImplementation = media::PlayerImplementation;
+
+    if (m_player == impl) return;
+
+    if (m_player) {
+        m_player->disconnect(this);
+    }
+
+    m_player = impl;
+    if (impl) {
+        QObject::connect(impl, &PlayerImplementation::seekedTo,
+                         this, &PlayerAdaptor::Seeked);
+
+        /* Property signals */
+        QObject::connect(impl, &PlayerImplementation::mprisPropertiesChanged,
+                         this, [this]() {
+            m_propertyNotifier.notify({
+                "CanPlay", "CanPause", "CanSeek",
+                "CanGoPrevious", "CanGoNext",
+            });
+        });
+        QObject::connect(impl, &PlayerImplementation::playbackStatusChanged,
+                         this, [this]() {
+            m_propertyNotifier.notify({ "PlaybackStatus" });
+        });
+        QObject::connect(impl, &PlayerImplementation::metadataForCurrentTrackChanged,
```

Comment:
I see, but I'd rather address these issues separately. This has mostly been a rewrite, and I've tried hard to preserve the logic in the tracklist code even though I found it weird and overcomplicated. Since it's not a regression, I'd leave it for later.